### PR TITLE
inherent gpio methods

### PR DESCRIPTION
**Pull request by burrbull**
See https://github.com/esp-rs/esp-hal/pull/1273#issuecomment-1995219928

### CODE REVIEW

**Comment by jessebraham** [comment]
Thanks for the PR.

We're still discussing what we're going to do regarding the various `embedded-hal` traits, so I'm hesitant to start refactoring our drivers specifically to accommodate them. Assuming that `embedded-hal-compat` gets updated, I'm leaning towards dropping the `embedded-hal@0.2.x` trait implementations entirely, though there still needs to be some discussion surrounding this.

I think given our current implementations that this PR is completely reasonable, however I think I would prefer to come to a decision regarding `embedded-hal` before committing to refactoring a bunch of drivers.

**Comment by MabezDev** [comment]
Maybe we could split this PR into two bits, the inherent infallible methods on the GPIO structs and the eh trait refactoring? Imo the inherent infallible methods we can accept now, before figuring out what to do with the traits.

**Comment by bjoernQ** [comment]
I like the idea of putting the hal implementations into modules

**Comment by burrbull** [comment]
Rebased.

**Comment by jessebraham** [comment]
> I like the idea of putting the hal implementations into modules

I very much don't, but guess I won't be too stubborn about it. I'd like some guarantee that _all_ of the other modules implementing these traits will be updated, because otherwise I have a feeling I will have to do this myself.

**Comment by bjoernQ** [comment]
> I like the idea of putting the hal implementations into modules

I personally like to have it in modules - I don't like to have it in different files t.b.h.

**Comment by jessebraham** [comment]
> Is problem in `mod`'s or different files?

At the end of the day it's largely just a matter of consistency, I guess my personal preferences don't matter as much (I seem to be in the minority as is often the case, anyway).

I've been putting a lot of work into trying to make our code base more consistent (this is still ongoing), and this feels like a step backwards unless these changes are made for _all_ applicable modules. I think it's reasonable for contributors to expect our drivers to be structured in similar ways, and I can see having only one or two driver like this causing confusion (or, we'll have to explain to contributors constantly what is expected in this regard).

Given the above, it follows that we should also split out modules for all other third-party trait packages, and this seems tedious and like a lot of work for essentially zero tangible gain.

But, I won't argue any more about this I suppose. Just my $0.02

**Comment by burrbull** [comment]
Is now better?

**Review by MabezDev — APPROVED** [review]
From my perspective, this LGTM. I'm a big fan of the infallible inherent methods, and I'm indifferent about the module layout, though I think cfg gated modules in the same file is a nice compromise between the two approaches.

**Review by bjoernQ — APPROVED** [review]
LGTM

File: examples/src/bin/ledc.rs [hunk 1]
```diff
@@ -30,7 +30,7 @@ fn main() -> ! {
     let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
 
     let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
-    let led = io.pins.gpio0.into_push_pull_output();
+    let led = io.pins.gpio8.into_push_pull_output();
```
Comment (MabezDev) [inline examples/src/bin/ledc.rs:33]: Did you mean to change these pins?

**Review by MabezDev — CHANGES_REQUESTED** [review]

File: examples/src/bin/ledc.rs (hunk 1 above)
Comment (burrbull) [reply examples/src/bin/ledc.rs]: No. Sorry. Fixed.

**Review by MabezDev — APPROVED** [review]
Thanks!

**Pull request merged**